### PR TITLE
tracing: general tracing support

### PR DESCRIPTION
**Pull request by wbpcode**
Signed-off-by: wbpcode <comems@msn.com>

Commit Message: tracing: general tracing support
Additional Description:

Check: #15784

Current tracing system only has great support to HTTP protocol. Other protocol such as Dubbo Proxy or Thrift Proxy cannot reuse tracers that designed for HTTP protocol.
This PR try to separate tracer implementation and HTTP protocol to make the current tracers protocol-independent.

There are mainly two protocol depentent design in current tracing system.
* First, `Http::RequestHeaderMap` for tracing context. Tracing Driver will try to obtain tracing context from downstream `Http::RequestHeaderMap` and create tracing span. Tracing span will inject new context to `Http::RequestHeaderMap` to transfer it to upstream.
* Second, Tracing driver factory will create HTTP tracing wrapper `HttpTracer` rather then raw `Driver`.

So, this PR design a new abstract class for tracing context to replace `Http::RequestHeaderMap` (While other protocols can implement an Http::RequestHeaderMap to directly reuse the existing Tracer, this requires a significant amount of work.):

```cpp

class TracingContext {
public:
  virtual ~TracingContext() = default;

  /**
   * Get tracing context value by key.
   *
   * @param key The context key of string view type. The context key should be a string view of a
   * const lowercase string.
   * @return The optional context value of string_view type.
   */
  virtual absl::optional<absl::string_view>
  getTracingContext(const absl::string_view key) const PURE;

  /**
   * Set new tracing context key/value pair.
   *
   * @param key The context key of string view type. The context key should be a string view of a
   * const lowercase string with a longer lifetime than the current TracingContext object.
   * @return The optional context value of string_view type.
   */
  virtual void setTracingContext(const absl::string_view key, const absl::string_view value) PURE;
};

```

And then, we update tracing driver factory to make it no deps to `HttpTracer`:
```cpp

class TracerFactory : public Config::TypedFactory {
public:
  ~TracerFactory() override = default;

  virtual Tracing::DriverSharedPtr createTracerDriver(const Protobuf::Message& config,
                                                      TracerFactoryContext& context) PURE;

```

All other code modifications are just to adapt to the above changes.

A more detailed doc:
https://docs.google.com/document/d/1gxk5vfHl-Gt3wiNZ-f46Dwmk2NFiEzFJQugmZmpSXk8/edit?usp=sharing

Risk Level: mid
Testing: [WIP]
Docs Changes: N/A
Release Notes: [WIP]

### CODE REVIEW

**Comment by wbpcode** [comment]
cc @mattklein123  I think this simple design can make tracers become protocol-independent. But I don’t know if the community has any opinions? 🤔

**Comment by wbpcode** [comment]
It's ready for a first review. @mattklein123

**Comment by lizan** [comment]
Anyone from @envoyproxy/first-pass-reviewers want to take this? At a quick glance it is in right direction.

**Review by rojkov — COMMENTED** [review]
Basically this looks very good to me. Just couple of things to clarify.

File: source/common/tracing/http_tracer_impl.cc
Comment (wbpcode) [reply source/common/tracing/http_tracer_impl.cc:320]: Get it.

File: source/common/tracing/http_tracer_impl.cc
Comment (wbpcode) [reply source/common/tracing/http_tracer_impl.cc:368]: 👌

File: source/common/tracing/http_tracer_impl.h
Comment (wbpcode) [reply source/common/tracing/http_tracer_impl.h:109]: Yep. This is a long-standing problem and I hope it can be resolved in the future.

```
  /**
   * Create a particular tracer driver implementation. If the implementation is unable to produce
   * an tracer driver with the provided parameters, it should throw an EnvoyException in the case
   * of general error or a Json::Exception if the json configuration is erroneous. The returned
   * pointer should always be valid.
   *
   * NOTE: Due to the corner case of OpenCensus, who can only support a single tracingYaroslav Skopets, 1 year ago: • tracing: ensure that OpenCensus is configured…
   *       configuration per entire process, the returned Driver instance is not guaranteed
   *       to be unique.
   *       That is why the return type has been changed to std::shared_ptr<> instead of a more
   *       idiomatic std::unique_ptr<>.
   *
   * @param config supplies the proto configuration for the HttpTracer
   * @param context supplies the factory context
   */
  virtual Tracing::DriverSharedPtr createTracerDriver(const Protobuf::Message& config,
                                                      TracerFactoryContext& context) PURE;
```

**Comment by wbpcode** [comment]
@rojkov Thank you very much for your kind review comments. I will update this PR this weekend. 😃

File: source/common/http/header_map_impl.h
Comment (rojkov) [inline source/common/http/header_map_impl.h:325]: Here we change the function's contract by adding a new implicit requirement for a caller that `key` must be lower case. Can we guarantee that or we don't care about casing now?

File: source/common/tracing/http_tracer_impl.h
Comment (rojkov) [reply source/common/tracing/http_tracer_impl.h:109]: I see. Thanks!

File: api/envoy/config/trace/v3/tracer.proto
Comment (wbpcode) [reply api/envoy/config/trace/v3/tracer.proto:23]: 👍

File: source/common/http/header_map_impl.h
Comment (wbpcode) [reply source/common/http/header_map_impl.h:325]: `getExisting(const absl::string_view key)` is a protected method that be used by `get(const LowerCaseString& key)` and `getTracingContext(const absl::string_view key)`. 

The former is the interface of HeaderMap and can always keep the case constraint. The latter is the interface of TracingContext, which is designed to not care about the case of the key.
When developers use TracingContext based on RequestHeaderMapImpl, they should ensure that the key meets the requirements of RequestHeaderMap (key must be lowercase).

Comment (wbpcode) [reply source/common/http/header_map_impl.h:325]: The design here is currently not perfect. I am thinking about whether it is possible to move Http::LowCaseString to common, and then TracingContext directly uses LowCaseString as the key type.

Comment (wbpcode) [reply source/common/http/header_map_impl.h:325]: In the design of TracingContext, I hope it is completely protocol-independent, so a simple string_view is used as the key type.
But in actual use, the combination of TracingContext and HeaderMap is not perfect.
In getTracingContext, I must assume that the content of key is a lowercase string. In setTracingContext, I must create a LowCaseString copy of the key.
I am thinking about whether I can directly move LowCaseString to common, and then TracingContext directly uses LowCaseString as the key type.
But I am still struggling with it. Maybe the community can give some opinions.

Comment (rojkov) [reply source/common/http/header_map_impl.h:325]: I don't have much experience with non-HTTP tracers and don't know if it's practical to convert to LowerCaseString always :( Will defer to the code owners to decide.

**Review by rojkov — DISMISSED** [review]
LGTM! @mattklein123 @lizan I think it's ready for a second pass.

File: api/envoy/config/trace/v3/tracer.proto
Comment (lizan) [inline api/envoy/config/trace/v3/tracer.proto:20]: Where this is used?

And the content of this message is exactly same as `TypedExtensionConfig` so we can just use that.

File: include/envoy/tracing/trace_context.h
Comment (lizan) [inline include/envoy/tracing/trace_context.h:29]: nit: no need top-level const in declarations.

```suggestion
  virtual absl::optional<absl::string_view> getTraceContext(absl::string_view key) const PURE;
```

File: include/envoy/tracing/trace_context.h
Comment (lizan) [inline include/envoy/tracing/trace_context.h:38]: ditto

File: api/envoy/config/trace/v3/tracer.proto
Comment (wbpcode) [reply api/envoy/config/trace/v3/tracer.proto:20]: It's designed to replace `envoy.config.trace.v3.Tracing.Http` in the future. There is no place to use it yet. But you are right, we can use `TypedExtensionConfig` directly.

File: include/envoy/common/lower_case_string.h
Comment (jmarantz) [inline include/envoy/common/lower_case_string.h:16]: drive-by comment: This is HTTP-specific and I think should under an http/ directory rather than under common.

I think I see why you wanted to do this: you wanted to add similar support for tracing that is not dependent on http, but then I think you would want to make the validator be pluggable or something, maybe via a template?

Comment (wbpcode) [reply include/envoy/common/lower_case_string.h:16]: Thanks.😄 Yes, I want add some similar support for tracing. A simple function pointer may be a good choice. Templates or virtual functions will introduce some type conversion problems, which is completely unnecessary. 
The validation logic of validHeaderString should be universal, that is, it does not contain newline characters and does not contain null. We can use it as the default validator.

File: source/common/http/header_map_impl.h
Comment (lizan) [inline source/common/http/header_map_impl.h:503]: We might want to just add an implicit conversion from `LowerCaseString` to `string_view`, so you don't need those overloads? Though we might want setTraceContext by reference or copy versions.

**Review by lizan — CHANGES_REQUESTED** [review]
In general, this PR involves many refactoring that can be split out to other PR so it would be easier to review.

File: include/envoy/common/lower_case_string.h
Comment (jmarantz) [reply include/envoy/common/lower_case_string.h]: I'm not understanding why templates would create type conversion problems. But if you are going to use a function pointer you should use a new-style lambda.

Also you probably want to avoid having include/envoy/common/... reference include/envoy/http/...  -- at least in my mind that should be a directed graph.

File: source/common/http/header_map_impl.h
Comment (wbpcode) [reply source/common/http/header_map_impl.h:503]: I agree with the sentiment that this PR needs to be split up. So if necessary, this can be continued in a new PR.

**Comment by wbpcode** [comment]
> In general, this PR involves many refactoring that can be split out to other PR so it would be easier to review.

@lizan @mattklein123 

This PR could be split out to four different PRs:
1. Remove the trace driver's dependency on `HttpTracerImpl`;
2. move `Http::LowerCaseString` to common for common use;
3. add TraceContext interface and implement it by RequestHeaderMapImpl;
4. replace RequestHeaderMap in tracers by TraceContext.

The first PR and fourth PR will involve a large number of file updates, but no logical changes.

File: include/envoy/common/lower_case_string.h
Comment (wbpcode) [reply include/envoy/common/lower_case_string.h]: > I'm not understanding why templates would create type conversion problems. But if you are going to use a function pointer you should use a new-style lambda.

Don't care, I should be wrong in my thinking.

> Also you probably want to avoid having include/envoy/common/... reference include/envoy/http/... -- at least in my mind that should be a directed graph.

🤔 I was thinking too. Perhaps we can simply treat validHeaderString as a general validator. It does not have any special HTTP-specific logic.

Comment (jmarantz) [reply include/envoy/common/lower_case_string.h]: It is based on http char-set, though, isn't it?

The use of a template, or for that matter, a function pointer, lambda, or pure virtual function call, is that you could put the generic impl into common/ and the http-specific variant in http/ and you have a clean separation.

Come to think of it, do you want to always lower-case names for non-http?

Comment (wbpcode) [reply include/envoy/common/lower_case_string.h]: @jmarantz Thank you very much for your patience and suggestions. 

If I just want to get a general LowerCaseString implementation, your idea is undoubtedly correct. But this question is somewhat different.

> Why I want a common LowerCaseString?

Because using RequestHeaderMapImpl to implement the `setTraceContext(absl::string_view)`/`getTraceContext(absl::string_view)` interface, additional overhead will be introduced. We need to convert absl::string_view to lower case and always use copy semantics. So I hope to design a general LowerCaseString and add two overloaded interfaces that use LowerCaseString to TraceContext. In this way, for existing Http Tracer users, the introduction of general tracing will not bring any additional overhead.

> Why is it difficult to achieve clean separation?

If I use virtual functions or inheritance solutions, use the base class in the TraceContext interface, such as LowerCaseStringBase. Then when RequestHeaderMapImpl implements the TraceContext interface, it will involve the protocol conversion from LowerCaseStringBase to Http::LowerCaseString (I now have some new ideals to circumvent this problem).
If I use function pointers, then we must use validHeaderString as the default value, because now Http::LowerCaseString has been widely used, we need to reduce the impact on existing code. The problem of using templates is similar to that of function pointers. In other words, LowerCaseString will depend on validHeaderString.

**Comment by wbpcode** [comment]
Check #16244 #16793 #17212
The main body of general tracing has been completed. **At present, the implementation of general tracing still has some rough points**, and some small PRs will be used to improve it in the future. Things to do next:

* Improve the TraceContext API and code quality.
* Extend dubbo/thrift proxy to verify the actual effect of general tracing. 

**Finally, thank you @lizan for your patience and help. **

**Pull request closed without merging**
